pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,118 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type process struct {
+	cmd     *exec.Cmd
+	running bool
+	lock    sync.RWMutex
+}
+
+func (p *process) kill() {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	if p.cmd.Process != nil {
+		p.cmd.Process.Kill()
+	}
+}
+
+func (p *process) getRunningState() bool {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	return p.running
+}
+
+func (p *process) setRunningState(state bool) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	p.running = state
+}
+
+func Start(ctx context.Context, httpsPort int, management *config.ScaledContext) error {
+	p := process{
+		lock: sync.RWMutex{},
+	}
+
+	// have two go routines running. One is to run telemetry if setting is true, one is to kill telemetry if setting is false
+	go func() {
+		ticker := time.NewTicker(time.Second * 5)
+		for {
+			select {
+			case <-ticker.C:
+				if settings.TelemetryOpt.Get() == "true" {
+					if !p.running {
+						token, err := createToken(management)
+						if err != nil {
+							logrus.Error(err)
+							continue
+						}
+						cmd := exec.Command("telemetry", "client", "--url", fmt.Sprintf("https://localhost:%d/v3", httpsPort), "--token-key", token)
+						cmd.Stdout = os.Stdout
+						cmd.Stderr = os.Stderr
+						p.cmd = cmd
+						if err := cmd.Start(); err != nil {
+							logrus.Error(err)
+							continue
+						}
+						p.setRunningState(true)
+						// in here we don't care the error, it will block until the telemetry process exited
+						cmd.Wait()
+						p.setRunningState(false)
+					}
+				}
+			case <-ctx.Done():
+				break
+			}
+		}
+	}()
+
+	go func() {
+		ticker := time.NewTicker(time.Second * 5)
+		for {
+			select {
+			case <-ticker.C:
+				if settings.TelemetryOpt.Get() == "false" {
+					if p.getRunningState() {
+						p.kill()
+						p.setRunningState(false)
+					}
+				}
+			case <-ctx.Done():
+				break
+			}
+		}
+	}()
+	return nil
+}
+
+func createToken(management *config.ScaledContext) (string, error) {
+	users, err := management.Management.Users("").List(metav1.ListOptions{
```

Comment:
think we can fetch users from cache, instead of fetching them from api directly?